Lazy uuid parsing

aztech-dev opened the pull request:
Improves performance when generating UUID's via the PECL UUID extension by delaying parsing of the uuid string until necessary.

Refs #56

jaymecd commented:
@aztech-dev, not sure if it's a good idea to create another class `LazyUuid`, which overwrites majority methods, as it may bring doubts for implementation use cases and especially when developer expects `LazyUuid` instance in his method by mistake.

Think it would be fine to implement this patch inside `Uuid` class directly.

aztech-dev commented:
Not too sure also, but not convinced Uuid is the right place either. Can you clarify your point about doubts for implementation ? I'm not sure I understand what you mean... Thanks :)

aztech-dev commented on src/LazyUuid.php:
```diff
@@ -0,0 +1,244 @@
+            return;
+        }
+
+        $this->fields = $this->codec->decode($this->uuidString)->getFieldsHex();
```
Missing assignment : $this->isParsed = true

jaymecd commented:
Ah, my bad :) Completely forgot about `UuidInterface`. 
Anyway, my explanation - normally you rely on a contract (interface), but `Uuid` is a class-as-a-contract (**before 3.0**). And now we bring `LazyUuid` class-as-a-contract.
~~Thats my problem~~
U could skip this comment, hard Monday :)

aztech-dev commented:
Ok :) Good thing I asked... but LazyUuid extends Uuid (most people are type-hinted against Uuid, not UuidInterface, so I felt it was safer) anyways, so it should not matter.

ramsey commented:
I'm also concerned about `LazyUuid`. Should we lazy-parse always for the performance gain, or are we only getting the performance benefit of lazy parsing when we're using `PeclUuid`?

aztech-dev commented:
I originally went for the 'always lazy parse' approach by returning a LazyUuid the `PeclUuidFactory::fromString()`, instead of returning a LazyUuid from the `uuid1()` and `uuid4()` methods. However, some unit tests for `fromString()` were failing since you can get by with passing an invalid/malformed string, so I did not keep the behavior. I'll try to push a smoother implementation in a bit.

ramsey commented:
I've done a bunch of merging around and making changes, so this is out of sync with master. You may want to rebase or submit a new PR.

ramsey commented:
@aztech-dev I'm getting ready to release a 3.0.0-alpha1. Is there any chance you can take a look and rebase this against master this week? If not, I can give it a try.

aztech-dev commented:
Sure, consider it done

aztech-dev commented:
I've rebased and fixed the (almost) minor issue that was laying around. However, I'm still not too sure about how to handle the lazy parsing. One hand, it could be nice to always lazy parse, on the other, lazy parsing skips validation of given string. Thoughts ?

aztech-dev commented:
I'm closing this in favor of #68 which feels like a better/cleaner option than this one.
